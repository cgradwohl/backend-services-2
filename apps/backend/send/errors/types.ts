import makeError from "make-error";
import { getErrorMessage } from "./lib";

// only status codes that indicate the service did not process the request should be retried

export enum SendErrorType {
  ALREADY_EXISTS = "ALREADY_EXISTS",
  FAILED_PRECONDITION = "FAILED_PRECONDITION",
  INTERNAL = "INTERNAL",
  INVALID_ARGUMENT = "INVALID_ARGUMENT",
  NOT_FOUND = "NOT_FOUND",
  RESOURCE_EXHAUSTED = "RESOURCE_EXHAUSTED",
  UNAVAILABLE = "UNAVAILABLE",
  UNKNOWN = "UNKNOWN",
}
type RetryableSendErrorType = Extract<
  SendErrorType,
  | SendErrorType.INTERNAL
  | SendErrorType.NOT_FOUND
  | SendErrorType.RESOURCE_EXHAUSTED
  | SendErrorType.UNAVAILABLE
  | SendErrorType.UNKNOWN
>;
type NonRetryableSendErrorType = Extract<
  SendErrorType,
  | SendErrorType.ALREADY_EXISTS
  | SendErrorType.FAILED_PRECONDITION
  | SendErrorType.INVALID_ARGUMENT
>;

const BaseError = makeError("BaseError");

export abstract class SendError extends BaseError {
  abstract retryable: boolean;
  abstract type: SendErrorType;
  public detail?: Record<string, string>;

  constructor(params: {
    error: unknown;
    type: SendErrorType;
    detail?: Record<string, string>;
  }) {
    super(getErrorMessage(params.error));

    if (params?.detail) {
      this.detail = params?.detail ?? (params?.error as any).detail;
    }

    if ((params.error as SendError)?.detail) {
      this.detail = {
        ...(params?.error as SendError)?.detail,
        ...params?.detail, // the last error to throw takes precedence
      };
    }
  }
}

export class NonRetryableSendError extends SendError {
  public retryable: false;
  public type: NonRetryableSendErrorType;

  constructor(params: {
    error: unknown;
    type: NonRetryableSendErrorType;
    detail?: Record<string, string>;
  }) {
    super(params);
    this.retryable = false;
    this.type = params.type;
  }
}

export class RetryableSendError extends SendError {
  public retryable: true;
  public type: RetryableSendErrorType;

  constructor(params: {
    error: unknown;
    type: RetryableSendErrorType;
    detail?: Record<string, string>;
  }) {
    super(params);
    this.retryable = true;
    this.type = params.type;
  }
}
