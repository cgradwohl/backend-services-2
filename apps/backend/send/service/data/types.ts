export type FilePathGenerator = (params: {
  id: string;
  name: SendDataEntity;
}) => `${string}/${string}_${string}.json`;

export interface SendDyanmoItem {
  created: string;
  updated: string;
  workspaceId: string;
}

export enum SendDataEntity {
  message = "message",
  request = "request",
  sequence = "sequence",
  sequenceAction = "sequence-action",
  scheduleJob = "schedule-job",
}
