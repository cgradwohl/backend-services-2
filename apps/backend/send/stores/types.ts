export type FilePathGenerator = ({
  requestId,
  messageId,
}: {
  requestId?: string;
  messageId?: string;
}) => string;

export type S3Get<T> = (payload: { filePath: string }) => Promise<T>;
export type S3Put<T> = (payload: {
  requestId?: string;
  messageId?: string;
  json: T;
}) => Promise<{ filePath: string }>;
