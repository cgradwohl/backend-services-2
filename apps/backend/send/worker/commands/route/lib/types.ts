import { IProviderConfiguration } from "~/send/types";

export type MatchingProvidersByChannel = {
  [channel: string]: IProviderConfiguration[];
};
