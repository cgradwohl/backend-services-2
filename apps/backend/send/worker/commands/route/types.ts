import { MessageStatusReason } from "~/lib/message-service/types";
import {
  IChannelConfiguration,
  IChannelProvider,
  IChannelProviderConfiguration,
  IConditionalConfig,
} from "~/types.api";

export interface IChannelSummary {
  channel: string;
  reason: string;
  selected: boolean;
}

export interface IUndeliverableEventLog {
  type: MessageStatusReason;
  reason: string;
  data: Record<string, unknown>;
}

export interface IRoutingSummary {
  channel: string;
  provider?: string;
  selected: boolean;
  timedout?: boolean;
  blockIds?: string[] | null;
  config?: IChannelConfiguration | IChannelProviderConfiguration | undefined;
  id?: string;
  reason?: string;
  providers?: IChannelProvider[] | null;
  taxonomy?: string;
  disabled?: boolean;
  label?: string;
  conditional?: IConditionalConfig;
  key?: string;
  // Provider Configuration ID
  configurationId?: string;
}
