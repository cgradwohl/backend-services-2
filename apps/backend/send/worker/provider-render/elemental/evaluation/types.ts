import { ElementalNodeIR } from "~/api/send/types";

export type ElementalRefs = { [refId: string]: ElementalNodeIR };
