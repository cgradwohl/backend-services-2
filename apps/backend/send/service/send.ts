import { nanoid } from "nanoid";
import { putRecord } from "~/lib/kinesis";
import { ISendProviderPayload, ISendRecord, SendService } from "../types";

const sendService: SendService = (tenantId: string) => {
  return {
    emit: async (payload: ISendProviderPayload) => {
      await putRecord<ISendRecord>({
        Data: {
          ...payload,
          tenantId,
          translated: payload?.translated ?? false,
        },
        PartitionKey: nanoid(),
        StreamName: process.env.PROVIDER_SEND_STREAM!,
      });
    },
  };
};

export default sendService;
