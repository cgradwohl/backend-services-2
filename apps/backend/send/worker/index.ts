import { createEventHandlerWithFailures } from "~/lib/kinesis/create-event-handler";
import { actionCommands } from "./commands";
import { IAction } from "../types";
import { CourierLogger } from "~/lib/logger";
import { NonRetryableSendError } from "../errors/types";
import { retryMessage } from "../utils/retry-message";
import { sendErrorMetric } from "~/lib/courier-emf/logger-metrics-utils";
const { logger } = new CourierLogger("ActionWorkerLogger");

async function worker(action: IAction) {
  try {
    const command = actionCommands[action.command];

    await command(action);
  } catch (error: unknown) {
    logger.error(`:: ${action?.command} COMMAND ERROR ::`, {
      action,
      error,
      tenantId: action?.tenantId,
      traceId: action?.requestId,
    });

    await sendErrorMetric({
      action,
      error,
      tenantId: action?.tenantId,
      traceId: action?.requestId,
    });

    // NOTE: This limits the scope of the new Error Retry Policy.
    // TODO: Remove after all command errors have been classified
    if (
      [
        "accept",
        "ad-hoc-list",
        "list",
        "list-pattern",
        "request",
        "route",
        "send-audiences",
        "send-audiences-member",
      ].includes(action?.command)
    ) {
      // an explicit non retryable error was thrown, so don't retry
      if (error instanceof NonRetryableSendError) {
        return;
      }

      await retryMessage({
        ...action,
        streamName: process.env.ACTION_STREAM!,
      });
    }
  }
}

export default createEventHandlerWithFailures<IAction>(
  worker,
  process.env.ACTION_STREAM_SEQUENCE_TABLE
);
