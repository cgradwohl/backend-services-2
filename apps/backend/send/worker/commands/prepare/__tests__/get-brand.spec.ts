import { MessageBrand, MessageChannels } from "~/api/send/types";
import {
  getBrand,
  getChannelBrandIds,
  getChannelBrands,
  getMessageBrands,
  messageBrandToIBrandContext,
  messageBrandV2ToIBrandContext,
} from "../get-brand";
import getBrandLegacy from "~/workers/lib/get-brand";
import getDefaultBrand from "~/workers/lib/get-default-brand";

const mockGetBrandLegacy = getBrandLegacy as jest.Mock;
const mockGetDefaultBrand = getDefaultBrand as jest.Mock;

jest.mock("~/workers/lib/get-brand");
jest.mock("~/workers/lib/get-default-brand");
jest.mock("~/workers/lib/get-latest-brand");
jest.mock("~/workers/lib/get-latest-default-brand");

describe("Get brand utils", () => {
  beforeEach(jest.resetAllMocks);

  describe("getBrand", () => {
    it("should call get default brand", async () => {
      const tenantId = "tenantId";

      await getBrand({ tenantId, scope: "published/production" });

      expect(mockGetDefaultBrand).toHaveBeenCalledWith(tenantId);
    });

    it("should call get brand legacy", async () => {
      const tenantId = "tenantId";
      const brandId = "brandId";

      await getBrand({ tenantId, scope: "published/production", brandId });

      expect(mockGetBrandLegacy).toHaveBeenCalledWith(tenantId, brandId, {
        extendDefaultBrand: true,
      });
    });
  });

  describe("getChannelBrandIds", () => {
    it("should return array of [channelName, brandId][]", () => {
      const channels: MessageChannels = {
        email: {
          brand_id: "brandId",
        },
        sms: {
          brand_id: "brandIdSms",
        },
      };

      const result = getChannelBrandIds(channels);
      expect(result).toEqual([
        ["email", "brandId"],
        ["sms", "brandIdSms"],
      ]);
    });
  });

  describe("getChannelBrands", () => {
    it("Should return a key value pair of channels and their IBrandContext", async () => {
      const channels: MessageChannels = {
        email: {
          brand_id: "brandId",
        },
        sms: {
          brand_id: "brandIdSms",
        },
      };
      mockGetBrandLegacy.mockResolvedValue({});

      await expect(
        getChannelBrands({
          channels,
          tenantId: "tenantId",
          scope: "published/production",
        })
      ).resolves.toEqual({
        email: {
          partials: {},
        },
        sms: {
          partials: {},
        },
      });
    });
  });

  describe("messageBrandV2ToIBrandContext", () => {
    it("should transform message brand to IBrandContext", () => {
      const inline: MessageBrand = {
        version: "2022-05-17",
        colors: {
          primary: "#ff0000",
          secondary: "#ff0000",
        },
        logo: {
          image: "https://example.com/logo.png",
          href: "https://example.com",
        },
        locales: {
          "eu-fr": {
            colors: {
              primary: "#eee",
              secondary: "#eee",
            },
            logo: {
              image: "https://fr.example.com/logo.png",
              href: "https://fr.example.com",
            },
          },
        },
      };
      const result = messageBrandV2ToIBrandContext(inline);
      delete result.created;
      delete result.updated;
      delete result.locales["eu-fr"].created;
      delete result.locales["eu-fr"].updated;
      expect(result).toMatchSnapshot();
    });
  });

  describe("messageBrandToIBrandContext", () => {
    it("should transform message brand to IBrandContext", () => {
      const inline: MessageBrand = {
        settings: {
          colors: {
            primary: "#ff0000",
            secondary: "#ff0000",
          },
          email: {
            header: {
              logo: {
                image: "https://example.com/logo.png",
                href: "https://example.com",
              },
            },
          },
        },
      };
      const result = messageBrandToIBrandContext(inline);
      delete result.created;
      delete result.updated;
      expect(result).toMatchSnapshot();
    });
  });

  describe("getMessageBrands", () => {
    it("should get the brands", async () => {
      expect.assertions(1);
      const channels: MessageChannels = {
        email: {
          brand_id: "brandId",
        },
        sms: {
          brand_id: "brandIdSms",
        },
      };
      mockGetBrandLegacy.mockResolvedValue({});

      await expect(
        getMessageBrands({
          channels,
          tenantId: "tenantId",
          scope: "published/production",
        })
      ).resolves.toEqual({
        main: undefined,
        channels: {
          email: {
            partials: {},
          },
          sms: {
            partials: {},
          },
        },
      });
    });
  });
});
