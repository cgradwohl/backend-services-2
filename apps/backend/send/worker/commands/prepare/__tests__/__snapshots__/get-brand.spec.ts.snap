// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Get brand utils messageBrandToIBrandContext should transform message brand to IBrandContext 1`] = `
Object {
  "creator": "inline",
  "id": "bdc0321cad38b7261d0196d61eaca8fa",
  "name": "inline",
  "settings": Object {
    "colors": Object {
      "primary": "#ff0000",
      "secondary": "#ff0000",
    },
    "email": Object {
      "header": Object {
        "logo": Object {
          "href": "https://example.com",
          "image": "https://example.com/logo.png",
        },
      },
    },
  },
  "updater": "inline",
  "version": "2020-06-19T18:51:36.083Z",
}
`;

exports[`Get brand utils messageBrandV2ToIBrandContext should transform message brand to IBrandContext 1`] = `
Object {
  "creator": "inline",
  "id": "02b63d04bdb30476273410d694ac6170",
  "locales": Object {
    "eu-fr": Object {
      "creator": "inline",
      "id": "7ca99416848a36c95bf6feea5050665e",
      "locales": Object {},
      "name": "inline",
      "settings": Object {
        "colors": Object {
          "primary": "#eee",
          "secondary": "#eee",
        },
        "email": Object {
          "header": Object {
            "logo": Object {
              "href": "https://fr.example.com",
              "image": "https://fr.example.com/logo.png",
            },
          },
        },
      },
      "updater": "inline",
      "version": "2022-05-17",
    },
  },
  "name": "inline",
  "settings": Object {
    "colors": Object {
      "primary": "#ff0000",
      "secondary": "#ff0000",
    },
    "email": Object {
      "header": Object {
        "logo": Object {
          "href": "https://example.com",
          "image": "https://example.com/logo.png",
        },
      },
    },
  },
  "updater": "inline",
  "version": "2022-05-17",
}
`;
